### ISSUE TITLE
Pulling images doesn't work in crio-1.18.0-rc1

### ISSUE THREAD

**Issue by lbogdan** [cri-o/cri-o#3612]
I get the following error when I try to pull an image using the [1.18.0-rc1 release](https://github.com/cri-o/cri-o/releases/download/v1.18.0-rc1/crio-v1.18.0-rc1.tar.gz):

```
$ crictl pull k8s.gcr.io/kube-apiserver:v1.18.2
FATA[0000] pulling image failed: rpc error: code = Unknown desc = error pinging docker registry k8s.gcr.io: Get "https://k8s.gcr.io/v2/": dial tcp: lookup k8s.gcr.io: device or resource busy
```

The `crio` log (set to `debug` loglevel) is as follows:

```
Apr 17 22:55:20 master-0 crio[13767]: time="2020-04-17 22:55:20.000049073Z" level=debug msg="Request: &PullImageRequest{Image:&ImageSpec{Image:k8s.gcr.io/kube-apiserver:v1.18.2,},Auth:nil,SandboxConfig:nil,}" file="go-grpc-middleware/chain.go:25" id=c4969608-3467-4e53-80d6-ca2239804364 name=/runtime.v1alpha2.ImageService/PullImage
Apr 17 22:55:20 master-0 crio[13767]: time="2020-04-17 22:55:20.000128846Z" level=debug msg="Skipping non-existing decryption_keys_path: /etc/crio/keys/" file="server/utils.go:262"
Apr 17 22:55:20 master-0 crio[13767]: time="2020-04-17 22:55:20.000439093Z" level=debug msg="Trying to access \"k8s.gcr.io/kube-apiserver:v1.18.2\"" file="docker/docker_image_src.go:68"
Apr 17 22:55:20 master-0 crio[13767]: time="2020-04-17 22:55:20.000492203Z" level=debug msg="Credentials not found" file="config/config.go:123"
Apr 17 22:55:20 master-0 crio[13767]: time="2020-04-17 22:55:20.000561864Z" level=debug msg="Using registries.d directory /etc/containers/registries.d for sigstore configuration" file="docker/lookaside.go:51"
Apr 17 22:55:20 master-0 crio[13767]: time="2020-04-17 22:55:20.000732186Z" level=debug msg=" Using \"default-docker\" configuration" file="docker/lookaside.go:169"
Apr 17 22:55:20 master-0 crio[13767]: time="2020-04-17 22:55:20.000764212Z" level=debug msg=" No signature storage configuration found for k8s.gcr.io/kube-apiserver:v1.18.2" file="docker/lookaside.go:174"
Apr 17 22:55:20 master-0 crio[13767]: time="2020-04-17 22:55:20.000794643Z" level=debug msg="Looking for TLS certificates and private keys in /etc/docker/certs.d/k8s.gcr.io" file="tlsclientconfig/tlsclientconfig.go:21"
Apr 17 22:55:20 master-0 crio[13767]: time="2020-04-17 22:55:20.000853022Z" level=debug msg="GET https://k8s.gcr.io/v2/" file="docker/docker_client.go:516"
Apr 17 22:55:20 master-0 crio[13767]: time="2020-04-17 22:55:20.001003283Z" level=debug msg="Ping https://k8s.gcr.io/v2/ err Get \"https://k8s.gcr.io/v2/\": dial tcp: lookup k8s.gcr.io: device or resource busy (&url.Error{Op:\"Get\", URL:\"https://k8s.gcr.io/v2/\", Err:(*net.OpError)(0xc0000f6b90)})" file="docker/docker_client.go:702"
Apr 17 22:55:20 master-0 crio[13767]: time="2020-04-17 22:55:20.001054407Z" level=debug msg="GET https://k8s.gcr.io/v1/_ping" file="docker/docker_client.go:516"
Apr 17 22:55:20 master-0 crio[13767]: time="2020-04-17 22:55:20.001148762Z" level=debug msg="Ping https://k8s.gcr.io/v1/_ping err Get \"https://k8s.gcr.io/v1/_ping\": dial tcp: lookup k8s.gcr.io: device or resource busy (&url.Error{Op:\"Get\", URL:\"https://k8s.gcr.io/v1/_ping\", Err:(*net.OpError)(0xc0000f6d20)})" file="docker/docker_client.go:729"
Apr 17 22:55:20 master-0 crio[13767]: time="2020-04-17 22:55:20.001188968Z" level=debug msg="Accessing \"k8s.gcr.io/kube-apiserver:v1.18.2\" failed: error pinging docker registry k8s.gcr.io: Get \"https://k8s.gcr.io/v2/\": dial tcp: lookup k8s.gcr.io: device or resource busy" file="docker/docker_image_src.go:73"
Apr 17 22:55:20 master-0 crio[13767]: time="2020-04-17 22:55:20.001228796Z" level=debug msg="Error preparing image k8s.gcr.io/kube-apiserver:v1.18.2: error pinging docker registry k8s.gcr.io: Get \"https://k8s.gcr.io/v2/\": dial tcp: lookup k8s.gcr.io: device or resource busy" file="server/image_pull.go:131" id=c4969608-3467-4e53-80d6-ca2239804364 name=/runtime.v1alpha2.ImageService/PullImage
Apr 17 22:55:20 master-0 crio[13767]: time="2020-04-17 22:55:20.001325296Z" level=debug msg="Response error: Get \"https://k8s.gcr.io/v2/\": dial tcp: lookup k8s.gcr.io: device or resource busy\nerror pinging docker registry k8s.gcr.io\ngithub.com/containers/image/v5/docker.(*dockerClient).detectPropertiesHelper\n\t/tmp/nix-build-cri-o-static.drv-0/go/src/github.com/cri-o/cri-o/vendor/github.com/containers/image/v5/docker/docker_client.go:720\ngithub.com/containers/image/v5/docker.(*dockerClient).detectProperties.func1\n\t/tmp/nix-build-cri-o-static.drv-0/go/src/github.com/cri-o/cri-o/vendor/github.com/containers/image/v5/docker/docker_client.go:753\nsync.(*Once).doSlow\n\t/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-go-1.14/share/go/src/sync/once.go:66\nsync.(*Once).Do\n\t/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-go-1.14/share/go/src/sync/once.go:57\ngithub.com/containers/image/v5/docker.(*dockerClient).detectProperties\n\t/tmp/nix-build-cri-o-static.drv-0/go/src/github.com/cri-o/cri-o/vendor/github.com/containers/image/v5/docker/docker_client.go:753\ngithub.com/containers/image/v5/docker.(*dockerClient).makeRequest\n\t/tmp/nix-build-cri-o-static.drv-0/go/src/github.com/cri-o/cri-o/vendor/github.com/containers/image/v5/docker/docker_client.go:421\ngithub.com/containers/image/v5/docker.(*dockerImageSource).fetchManifest\n\t/tmp/nix-build-cri-o-static.drv-0/go/src/github.com/cri-o/cri-o/vendor/github.com/containers/image/v5/docker/docker_image_src.go:188\ngithub.com/containers/image/v5/docker.(*dockerImageSource).ensureManifestIsLoaded\n\t/tmp/nix-build-cri-o-static.drv-0/go/src/github.com/cri-o/cri-o/vendor/github.com/containers/image/v5/docker/docker_image_src.go:221\ngithub.com/containers/image/v5/docker.newImageSourceAttempt\n\t/tmp/nix-build-cri-o-static.drv-0/go/src/github.com/cri-o/cri-o/vendor/github.com/containers/image/v5/docker/docker_image_src.go:126\ngithub.com/containers/image/v5/docker.newImageSource\n\t/tmp/nix-build-cri-o-static.drv-0/go/src/github.com/cri-o/cri-o/vendor/github.com/containers/image/v5/docker/docker_image_src.go:69\ngithub.com/containers/image/v5/docker.newImage\n\t/tmp/nix-build-cri-o-static.drv-0/go/src/github.com/cri-o/cri-o/vendor/github.com/containers/image/v5/docker/docker_image.go:27\ngithub.com/containers/image/v5/docker.dockerReference.NewImage\n\t/tmp/nix-build-cri-o-static.drv-0/go/src/github.com/cri-o/cri-o/vendor/github.com/containers/image/v5/docker/docker_transport.go:138\ngithub.com/cri-o/cri-o/internal/storage.(*imageService).PrepareImage\n\t/tmp/nix-build-cri-o-static.drv-0/go/src/github.com/cri-o/cri-o/internal/storage/image.go:370\ngithub.com/cri-o/cri-o/server.(*Server).pullImage\n\t/tmp/nix-build-cri-o-static.drv-0/go/src/github.com/cri-o/cri-o/server/image_pull.go:117\ngithub.com/cri-o/cri-o/server.(*Server).PullImage\n\t/tmp/nix-build-cri-o-static.drv-0/go/src/github.com/cri-o/cri-o/server/image_pull.go:76\nk8s.io/cri-api/pkg/apis/runtime/v1alpha2._ImageService_PullImage_Handler.func1\n\t/tmp/nix-build-cri-o-static.drv-0/go/src/github.com/cri-o/cri-o/vendor/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.pb.go:8335\ngithub.com/cri-o/cri-o/internal/log.UnaryInterceptor.func1\n\t/tmp/nix-build-cri-o-static.drv-0/go/src/github.com/cri-o/cri-o/internal/log/interceptors.go:61\ngithub.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1.1.1\n\t/tmp/nix-build-cri-o-static.drv-0/go/src/github.com/cri-o/cri-o/vendor/github.com/grpc-ecosystem/go-grpc-middleware/chain.go:25\ngithub.com/cri-o/cri-o/server/metrics.UnaryInterceptor.func1\n\t/tmp/nix-build-cri-o-static.drv-0/go/src/github.com/cri-o/cri-o/server/metrics/interceptors.go:24\ngithub.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1.1.1\n\t/tmp/nix-build-cri-o-static.drv-0/go/src/github.com/cri-o/cri-o/vendor/github.com/grpc-ecosystem/go-grpc-middleware/chain.go:25\ngithub.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1\n\t/tmp/nix-build-cri-o-static.drv-0/go/src/github.com/cri-o/cri-o/vendor/github.com/grpc-ecosystem/go-grpc-middleware/chain.go:34\nk8s.io/cri-api/pkg/apis/runtime/v1alpha2._ImageService_PullImage_Handler\n\t/tmp/nix-build-cri-o-static.drv-0/go/src/github.com/cri-o/cri-o/vendor/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.pb.go:8337\ngoogle.golang.org/grpc.(*Server).processUnaryRPC\n\t/tmp/nix-build-cri-o-static.drv-0/go/src/github.com/cri-o/cri-o/vendor/google.golang.org/grpc/server.go:1082\ngoogle.golang.org/grpc.(*Server).handleStream\n\t/tmp/nix-build-cri-o-static.drv-0/go/src/github.com/cri-o/cri-o/vendor/google.golang.org/grpc/server.go:1405\ngoogle.golang.org/grpc.(*Server).serveStreams.func1.1\n\t/tmp/nix-build-cri-o-static.drv-0/go/src/github.com/cri-o/cri-o/vendor/google.golang.org/grpc/server.go:746\nruntime.goexit\n\t/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-go-1.14/share/go/src/runtime/asm_amd64.s:1373" file="go-grpc-middleware/chain.go:25" id=c4969608-3467-4e53-80d6-ca2239804364 name=/runtime.v1alpha2.ImageService/PullImage
```

[crio 1.17.2~1 Ubuntu 18.04 package](http://download.opensuse.org/repositories/devel:/kubic:/libcontainers:/stable/xUbuntu_18.04/amd64) works as expected.

**Comment by lbogdan** [comment]
I did a bit of researching, and it seems to be related to how the Go `net` package handles [name resolution](https://golang.org/pkg/net/#hdr-Name_Resolution), specifically it seems that the cgo resolver errors with "device or resource busy" for some reason.

As a workaround, forcing the pure Go resolver by starting `crio` with `GODEBUG=netdns=go` works.

**Comment by saschagrunert** [comment]
Hey @lbogdan, thank you for the detailed debugging. I assume you've built the binary locally or did you use the static bundle?

**Comment by lbogdan** [comment]
Using the static 1.18.0-rc1 build from https://github.com/cri-o/cri-o/releases/download/v1.18.0-rc1/crio-v1.18.0-rc1.tar.gz .

**Comment by saschagrunert** [comment]
We got some warnings in the glibc build process which did not occur on musl:
```
GO111MODULE=on go build --mod=vendor  -ldflags '-s -w -X github.com/cri-o/cri-o/internal/pkg/criocli.DefaultsPath="" -X github.com/cri-o/cri-o/internal/version.buildDate=2020-04-18T03:29:08Z -X github.com/cri-o/cri-o/internal/version.gitCommit=2a9565bb39dc4d9ea19fabdb775f74350c3cd965 -X github.com/cri-o/cri-o/internal/version.gitTreeState=clean -X github.com/cri-o/cri-o/internal/version.version=1.18.0-dev -linkmode external -extldflags "-static -lm"' -tags "apparmor seccomp selinux containers_image_ostree_stub" -o bin/crio-status github.com/cri-o/cri-o/cmd/crio-status
# github.com/cri-o/cri-o/cmd/crio-status
/nix/store/nfs9wf6ipffr9g9ylzb4i9c1i607nxvq-binutils-2.31.1/bin/ld: /tmp/nix-build-cri-o-static.drv-0/go-link-789894549/000022.o: in function `mygetgrouplist':
/nix/store/nha5hv6zzdixkdp6icz8319ylk0ii60n-go-1.14/share/go/src/os/user/getgrouplist_unix.go:16: warning: Using 'getgrouplist' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
/nix/store/nfs9wf6ipffr9g9ylzb4i9c1i607nxvq-binutils-2.31.1/bin/ld: /tmp/nix-build-cri-o-static.drv-0/go-link-789894549/000021.o: in function `mygetgrgid_r':
/nix/store/nha5hv6zzdixkdp6icz8319ylk0ii60n-go-1.14/share/go/src/os/user/cgo_lookup_unix.go:38: warning: Using 'getgrgid_r' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
/nix/store/nfs9wf6ipffr9g9ylzb4i9c1i607nxvq-binutils-2.31.1/bin/ld: /tmp/nix-build-cri-o-static.drv-0/go-link-789894549/000021.o: in function `mygetgrnam_r':
/nix/store/nha5hv6zzdixkdp6icz8319ylk0ii60n-go-1.14/share/go/src/os/user/cgo_lookup_unix.go:43: warning: Using 'getgrnam_r' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
/nix/store/nfs9wf6ipffr9g9ylzb4i9c1i607nxvq-binutils-2.31.1/bin/ld: /tmp/nix-build-cri-o-static.drv-0/go-link-789894549/000021.o: in function `mygetpwnam_r':
/nix/store/nha5hv6zzdixkdp6icz8319ylk0ii60n-go-1.14/share/go/src/os/user/cgo_lookup_unix.go:33: warning: Using 'getpwnam_r' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
/nix/store/nfs9wf6ipffr9g9ylzb4i9c1i607nxvq-binutils-2.31.1/bin/ld: /tmp/nix-build-cri-o-static.drv-0/go-link-789894549/000021.o: in function `mygetpwuid_r':
/nix/store/nha5hv6zzdixkdp6icz8319ylk0ii60n-go-1.14/share/go/src/os/user/cgo_lookup_unix.go:28: warning: Using 'getpwuid_r' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
/nix/store/nfs9wf6ipffr9g9ylzb4i9c1i607nxvq-binutils-2.31.1/bin/ld: /tmp/nix-build-cri-o-static.drv-0/go-link-789894549/000004.o: in function `_cgo_26061493d47f_C2func_getaddrinfo':
/tmp/go-build/cgo-gcc-prolog:58: warning: Using 'getaddrinfo' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
```

Is there a way to enforce `GODEBUG=netdns=go` during the build process?

**Comment by lbogdan** [comment]
From https://www.osso.nl/blog/golang-statically-linked/ :

> TL;DR: Use `CGO_ENABLED=0` or `-tags netgo` to create a static executable.

**Comment by saschagrunert** [comment]
> From https://www.osso.nl/blog/golang-statically-linked/ :
> 
> > TL;DR: Use `CGO_ENABLED=0` or `-tags netgo` to create a static executable.

Thank you, I'll spin-up a PR to fix this.
